### ISSUE TITLE
Optimistic timeout action must throw OCE for proper signaling; TimeoutPolicy should not mask user exception

### ISSUE THREAD

**Issue by sandersaares** [App-vNext/Polly#620]
I was confused to find that my timeout policy was not throwing a `TimeoutRejectedException` on timeout.

```csharp
static void Dingus(CancellationToken ct)
{
    Console.WriteLine($"It's 12 o'clock and all is well");

    // Just wait until cancelled.
    ct.WaitHandle.WaitOne();
}
```

It appears that using the optimistic strategy, it is necessary to actually throw `OperationCancelledException` (e.g. using `ct.ThrowIfCancellationRequested();`) in order to cause Polly to throw `TimeoutRejectedException`. This makes sense but is not documented.

I request this be documented.

**Edit:** I discover that throwing any exception is enough to trigger `TimeoutRejectedException` - it does not have to be `OperationCancelledException`. Another unexpected point that I request be documented.

**Comment by reisenberger** [comment]
Thanks @sandersaares .  Yes, optimistic strategy expects executed delegates to express cancellation by throwing `OperationCanceledException`. We've updated the [wiki documentation](https://github.com/App-vNext/Polly/wiki/Timeout#optimistic-timeout) to make that clearer. 

Re:

>throwing any exception is enough to trigger TimeoutRejectedException [EDIT: if the timeout has also occurred...]- it does not have to be OperationCancelledException. 

I'm going to consider treating this as an edge-case that we might squash out (consider a minor edge case bug) rather than document.  

Provided that code does follow the recommended pattern (`ct.ThrowIfCancellationRequested();`), this should be a rare edge-case. To occur, it would require a race condition whereby cancellation occurs in the same 'instant' (during the time that) the runtime is organising throwing the non-cancellation exception back on to the policy. However, in that eventuality, I confirm the code is preferring to express timeout over the thrown exception, which seems undesirable.  It should be rare. But I'm considering squashing it out.

**Comment by reisenberger** [comment]
Closing. #725 squashes out the edge-case issue that a timeout could (in rare race circumstances) mask a user-exception. TimeoutRejectedExecption rather than the user exception would previously be thrown by TimeoutPolicy in the rare case that a timeout occurred while a user-exception was being marshalled into/through TimeoutPolicy code.  #725 fixes this so that original user exception is still rethrown in this case.

Merged to dev branch - will be released when v7.2.0 released.
